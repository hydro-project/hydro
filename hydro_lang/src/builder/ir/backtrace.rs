//! Platform-independent interface for collecting backtraces, used in the Hydro IR to
//! trace the origin of each node.

#[cfg(feature = "build")]
use std::cell::RefCell;
#[cfg(feature = "build")]
use std::fmt::Debug;

#[cfg(not(feature = "build"))]
/// A dummy backtrace element with no data. Enable the `build` feature to collect backtraces.
#[derive(Clone)]
pub struct Backtrace;

#[cfg(feature = "build")]
/// Captures an entire backtrace, whose elements will be lazily resolved. See
/// [`Backtrace::elements`] for more information.
#[derive(Clone)]
pub struct Backtrace {
    skip_count: usize,
    inner: RefCell<backtrace::Backtrace>,
    resolved: RefCell<Option<Vec<BacktraceElement>>>,
}

impl Backtrace {
    #[cfg(feature = "build")]
    #[inline(never)]
    pub(crate) fn get_backtrace(skip_count: usize) -> Backtrace {
        let backtrace = backtrace::Backtrace::new_unresolved();
        Backtrace {
            skip_count,
            inner: RefCell::new(backtrace),
            resolved: RefCell::new(None),
        }
    }

    #[cfg(not(feature = "build"))]
    pub(crate) fn get_backtrace(_skip_count: usize) -> Backtrace {
        panic!();
    }

    #[cfg(feature = "build")]
    /// Gets the elements of the backtrace including inlined frames.
    ///
    /// Excludes all backtrace elements up to the original `get_backtrace` call as
    /// well as additional skipped frames from that call. Also drops the suffix
    /// of frames from `__rust_begin_short_backtrace` onwards.
    pub fn elements(&self) -> Vec<BacktraceElement> {
        self.resolved
            .borrow_mut()
            .get_or_insert_with(|| {
                let mut inner_borrow = self.inner.borrow_mut();
                inner_borrow.resolve();
                inner_borrow
                    .frames()
                    .iter()
                    .skip_while(|f| {
                        !(f.symbol_address() as usize == Backtrace::get_backtrace as usize
                            || f.symbols()
                                .first()
                                .and_then(|s| s.name())
                                .and_then(|n| n.as_str())
                                .is_some_and(|n| n.contains("get_backtrace")))
                    })
                    .skip(1)
                    .take_while(|f| {
                        !f.symbols()
                            .last()
                            .and_then(|s| s.name())
                            .and_then(|n| n.as_str())
                            .is_some_and(|n| n.contains("__rust_begin_short_backtrace"))
                    })
                    .flat_map(|frame| frame.symbols())
                    .skip(self.skip_count)
                    .map(|symbol| {
                        let full_fn_name = symbol.name().unwrap().to_string();
                        BacktraceElement {
                            fn_name: full_fn_name
                                .rfind("::")
                                .map(|idx| full_fn_name.split_at(idx).0.to_string())
                                .unwrap_or(full_fn_name),
                            filename: symbol.filename().map(|f| f.display().to_string()),
                            lineno: symbol.lineno(),
                            colno: symbol.colno(),
                            addr: symbol.addr().map(|a| a as usize),
                        }
                    })
                    .collect()
            })
            .clone()
    }
}

#[cfg(feature = "build")]
/// A single frame of a backtrace, corresponding to a single function call.
#[derive(Clone)]
pub struct BacktraceElement {
    /// The name of the function that was called.
    pub fn_name: String,
    /// The path to the file where this call occured.
    pub filename: Option<String>,
    /// The line number of the function call.
    pub lineno: Option<u32>,
    /// The column number of the function call.
    pub colno: Option<u32>,
    /// The address of the instruction corresponding to this function call.
    pub addr: Option<usize>,
}

#[cfg(feature = "build")]
impl Debug for BacktraceElement {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // filename / addr is unstable across platforms so we drop it
        f.debug_struct("BacktraceElement")
            .field("fn_name", &self.fn_name)
            .field("lineno", &self.lineno)
            .field("colno", &self.colno)
            .finish()
    }
}

#[cfg(test)]
mod tests {
    #[cfg(unix)]
    use super::*;

    #[cfg(unix)]
    #[test]
    fn test_backtrace() {
        let backtrace = Backtrace::get_backtrace(0);
        let elements = backtrace.elements();

        hydro_build_utils::assert_debug_snapshot!(elements);
    }
}
