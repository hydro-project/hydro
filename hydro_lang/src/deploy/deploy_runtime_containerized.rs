#![allow(
    unused,
    reason = "unused in trybuild but the __staged version is needed"
)]
#![allow(missing_docs, reason = "used internally")]

use std::collections::HashMap;
use std::future::Future;
use std::net::SocketAddr;
use std::ops::{Deref, DerefMut};
use std::pin::Pin;
use std::sync::Arc;
use std::task::{Context, Poll};

use bytes::BytesMut;
use futures::{FutureExt, Sink, SinkExt, Stream, StreamExt};
use sinktools::demux_map_lazy::LazyDemuxSink;
use sinktools::lazy::{LazySink, LazySource};
use stageleft::{QuotedWithContext, q};
use tokio::net::{TcpListener, TcpStream};
use tokio_util::codec::{FramedRead, FramedWrite, LengthDelimitedCodec};

use crate::location::MemberId;

pub fn deploy_containerized_o2o(target: &str, bind_addr: &str) -> (syn::Expr, syn::Expr) {
    (
        q!(LazySink::<_, _, _, bytes::Bytes>::new(move || Box::pin(
            async move {
                eprintln!("CONNECTING TO: {}", target);
                Result::<_, std::io::Error>::Ok(FramedWrite::new(
                    TcpStream::connect(target).await?,
                    LengthDelimitedCodec::new(),
                ))
            }
        )))
        .splice_untyped_ctx(&()),
        q!(LazySource::new(move || Box::pin(async move {
            let listener = TcpListener::bind(bind_addr).await?;
            let (stream, peer) = listener.accept().await?;
            eprintln!("ACCEPTED FROM: {peer:?}");
            Result::<_, std::io::Error>::Ok(FramedRead::new(stream, LengthDelimitedCodec::new()))
        })))
        .splice_untyped_ctx(&()),
    )
}

pub fn deploy_containerized_o2m<Tag: Unpin>(port: u16, bind_addr: &str) -> (syn::Expr, syn::Expr) {
    (
        QuotedWithContext::<'static, LazyDemuxSink<MemberId<Tag>, _, _>, ()>::splice_untyped_ctx(
            q!(sinktools::demux_map_lazy::<_, _, _, _>(
                move |key: MemberId<_>| {
                    LazySink::<_, _, _, bytes::Bytes>::new(move || {
                        Box::pin(async move {
                            eprintln!("CONNECTING TO: {}:{}", key.get_container_name(), port);
                            let mut sink = FramedWrite::new(
                                TcpStream::connect(format!(
                                    "{}:{}",
                                    key.get_container_name(),
                                    port
                                ))
                                .await?,
                                LengthDelimitedCodec::new(),
                            );

                            // sink.send(bytes::Bytes::copy_from_slice(
                            //     bincode::serialize(&std::env::var("CONTAINER_NAME").unwrap())
                            //         .unwrap()
                            //         .as_slice(),
                            // ))
                            // .await?;

                            Result::<_, std::io::Error>::Ok(sink)
                        })
                    })
                }
            )),
            &(),
        ),
        q!(LazySource::new(move || Box::pin(async move {
            let listener = TcpListener::bind(bind_addr).await?;
            let (stream, peer) = listener.accept().await?;
            eprintln!("ACCEPTED FROM: {peer}");

            Result::<_, std::io::Error>::Ok(FramedRead::new(stream, LengthDelimitedCodec::new()))
        })))
        .splice_untyped_ctx(&()),
    )
}

pub fn deploy_containerized_m2o<'a, Tag>(
    target: &str,
    bind_addr: &'a str,
) -> (syn::Expr, syn::Expr) {
    (
        q!(LazySink::<_, _, _, bytes::Bytes>::new(move || {
            Box::pin(async move {
                eprintln!("CONNECTING TO: {}", target);

                let mut sink = FramedWrite::new(
                    TcpStream::connect(target).await?,
                    LengthDelimitedCodec::new(),
                );

                sink.send(bytes::Bytes::copy_from_slice(
                    bincode::serialize(&std::env::var("CONTAINER_NAME").unwrap())
                        .unwrap()
                        .as_slice(),
                ))
                .await?;

                Result::<_, std::io::Error>::Ok(sink)
            })
        }))
        .splice_untyped_ctx(&()),
        QuotedWithContext::<'a, LazySource<_, _, _, Result<(MemberId<Tag>, BytesMut), _>>, ()>::splice_untyped_ctx(
            q!(LazySource::new(move || Box::pin(async move {
                let listener = TcpListener::bind(bind_addr).await?;
                Result::<_, std::io::Error>::Ok(
                    futures::stream::unfold(listener, |listener| {
                        Box::pin(async move {
                            let (stream, peer) = listener.accept().await.ok()?;
                            let mut source = FramedRead::new(stream, LengthDelimitedCodec::new());
                            let from =
                                bincode::deserialize::<String>(&source.next().await?.ok()?[..])
                                    .ok()?;

                            eprintln!("ACCEPTING FROM: {} - {}", from, peer);

                            Some((
                                source.map(move |v| {
                                    v.map(|v| (MemberId::from_container_name(from.clone()), v))
                                }),
                                listener,
                            ))
                        })
                    })
                    .flatten_unordered(None),
                )
            }))),
            &(),
        ),
    )
}

pub fn deploy_containerized_e2o(bind_addr: &str) -> syn::Expr {
    q!(LazySource::new(move || Box::pin(async move {
        let listener = TcpListener::bind(bind_addr).await?;
        let (stream, peer) = listener.accept().await?;
        eprintln!("ACCEPTED FROM: {peer:?}");
        Result::<_, std::io::Error>::Ok(FramedRead::new(stream, LengthDelimitedCodec::new()))
    })))
    .splice_untyped_ctx(&())
}

pub fn deploy_containerized_o2e(target: &str) -> syn::Expr {
    q!(LazySink::<_, _, _, bytes::Bytes>::new(move || Box::pin(
        async move {
            eprintln!("CONNECTING TO: {}", target);

            Result::<_, std::io::Error>::Ok(FramedWrite::new(
                TcpStream::connect(target).await?,
                LengthDelimitedCodec::new(),
            ))
        }
    )))
    .splice_untyped_ctx(&())
}
