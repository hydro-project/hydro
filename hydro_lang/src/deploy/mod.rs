//! Infrastructure for deploying Hydro programs to the cloud using [`hydro_deploy`].

#[cfg(feature = "deploy_integration")]
mod deploy_runtime;

#[cfg(stageleft_runtime)]
#[cfg(feature = "deploy")]
pub(crate) mod trybuild;

#[cfg(stageleft_runtime)]
#[cfg(feature = "deploy")]
mod trybuild_rewriters;

#[cfg(stageleft_runtime)]
#[cfg(feature = "deploy")]
#[cfg_attr(docsrs, doc(cfg(feature = "deploy")))]
pub use trybuild::init_test;

#[cfg(stageleft_runtime)]
#[cfg(feature = "deploy")]
#[cfg_attr(docsrs, doc(cfg(feature = "deploy")))]
pub mod deploy_graph;

#[cfg(stageleft_runtime)]
#[cfg(feature = "deploy")]
#[cfg_attr(docsrs, doc(cfg(feature = "deploy")))]
pub use deploy_graph::*;
