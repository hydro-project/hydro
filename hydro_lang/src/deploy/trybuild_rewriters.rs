use syn::visit_mut::VisitMut;

pub struct ReplaceCrateNameWithStaged {
    pub crate_name: String,
    pub is_test: bool,
}

impl VisitMut for ReplaceCrateNameWithStaged {
    fn visit_type_path_mut(&mut self, i: &mut syn::TypePath) {
        if let Some(first) = i.path.segments.first() {
            if first.ident == self.crate_name {
                let tail = i.path.segments.iter().skip(1).collect::<Vec<_>>();

                if self.is_test {
                    *i = syn::parse_quote!(crate::__staged #(::#tail)*);
                } else {
                    let crate_ident = syn::Ident::new(&self.crate_name, first.ident.span());
                    *i = syn::parse_quote!(#crate_ident::__staged #(::#tail)*);
                }
            }
        }

        syn::visit_mut::visit_type_path_mut(self, i);
    }

    fn visit_use_path_mut(&mut self, i: &mut syn::UsePath) {
        if i.ident == "crate" && !self.is_test {
            i.ident = syn::Ident::new(&self.crate_name, i.ident.span());
        }
    }
}
