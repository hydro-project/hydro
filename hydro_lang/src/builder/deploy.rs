use std::collections::{BTreeMap, HashMap};
use std::io::Error;
use std::marker::PhantomData;
use std::pin::Pin;

use dfir_rs::bytes::Bytes;
use dfir_rs::futures::{Sink, Stream};
use proc_macro2::Span;
use serde::Serialize;
use serde::de::DeserializeOwned;
use stageleft::QuotedWithContext;

use super::built::build_inner;
use super::compiled::CompiledFlow;
use crate::deploy::{
    ClusterSpec, Deploy, ExternalSpec, IntoProcessSpec, LocalDeploy, Node, ProcessSpec,
    RegisterPort,
};
use crate::ir::HydroLeaf;
use crate::location::external_process::{
    ExternalBincodeSink, ExternalBincodeStream, ExternalBytesPort,
};
use crate::location::{Cluster, ExternalProcess, Location, LocationId, Process};
use crate::staging_util::Invariant;

pub struct DeployFlow<'a, D: LocalDeploy<'a>> {
    pub(super) ir: Vec<HydroLeaf>,

    /// Deployed instances of each process in the flow
    pub(super) processes: HashMap<usize, D::Process>,

    /// Lists all the processes that were created in the flow, same ID as `processes`
    /// but with the type name of the tag.
    pub(super) process_id_name: Vec<(usize, String)>,

    pub(super) externals: HashMap<usize, D::ExternalProcess>,
    pub(super) external_id_name: Vec<(usize, String)>,

    pub(super) clusters: HashMap<usize, D::Cluster>,
    pub(super) cluster_id_name: Vec<(usize, String)>,
    pub(super) used: bool,

    pub(super) _phantom: Invariant<'a, D>,
}

impl<'a, D: LocalDeploy<'a>> Drop for DeployFlow<'a, D> {
    fn drop(&mut self) {
        if !self.used {
            panic!(
                "Dropped DeployFlow without instantiating, you may have forgotten to call `compile` or `deploy`."
            );
        }
    }
}

impl<'a, D: LocalDeploy<'a>> DeployFlow<'a, D> {
    pub fn ir(&self) -> &Vec<HydroLeaf> {
        &self.ir
    }

    pub fn with_process<P>(
        mut self,
        process: &Process<P>,
        spec: impl IntoProcessSpec<'a, D>,
    ) -> Self {
        let tag_name = std::any::type_name::<P>().to_string();
        self.processes.insert(
            process.id,
            spec.into_process_spec().build(process.id, &tag_name),
        );
        self
    }

    pub fn with_remaining_processes<S: IntoProcessSpec<'a, D> + 'a>(
        mut self,
        spec: impl Fn() -> S,
    ) -> Self {
        for (id, name) in &self.process_id_name {
            self.processes
                .insert(*id, spec().into_process_spec().build(*id, name));
        }

        self
    }

    pub fn with_external<P>(
        mut self,
        process: &ExternalProcess<P>,
        spec: impl ExternalSpec<'a, D>,
    ) -> Self {
        let tag_name = std::any::type_name::<P>().to_string();
        self.externals
            .insert(process.id, spec.build(process.id, &tag_name));
        self
    }

    pub fn with_remaining_externals<S: ExternalSpec<'a, D> + 'a>(
        mut self,
        spec: impl Fn() -> S,
    ) -> Self {
        for (id, name) in &self.external_id_name {
            self.externals.insert(*id, spec().build(*id, name));
        }

        self
    }

    pub fn with_cluster<C>(mut self, cluster: &Cluster<C>, spec: impl ClusterSpec<'a, D>) -> Self {
        let tag_name = std::any::type_name::<C>().to_string();
        self.clusters
            .insert(cluster.id, spec.build(cluster.id, &tag_name));
        self
    }

    pub fn with_remaining_clusters<S: ClusterSpec<'a, D> + 'a>(
        mut self,
        spec: impl Fn() -> S,
    ) -> Self {
        for (id, name) in &self.cluster_id_name {
            self.clusters.insert(*id, spec().build(*id, name));
        }

        self
    }

    pub fn compile_no_network(mut self) -> CompiledFlow<'a, D::GraphId> {
        self.used = true;

        CompiledFlow {
            dfir: build_inner(&mut self.ir),
            extra_stmts: BTreeMap::new(),
            _phantom: PhantomData,
        }
    }
}

impl<'a, D: Deploy<'a>> DeployFlow<'a, D> {
    pub fn compile(mut self, env: &D::CompileEnv) -> CompiledFlow<'a, D::GraphId> {
        self.used = true;

        let mut seen_tees: HashMap<_, _> = HashMap::new();
        let mut seen_tee_locations: HashMap<_, _> = HashMap::new();
        self.ir.iter_mut().for_each(|leaf| {
            leaf.compile_network::<D>(
                env,
                &mut seen_tees,
                &mut seen_tee_locations,
                &self.processes,
                &self.clusters,
                &self.externals,
            );
        });

        let extra_stmts = self.extra_stmts(env);

        CompiledFlow {
            dfir: build_inner(&mut self.ir),
            extra_stmts,
            _phantom: PhantomData,
        }
    }

    fn extra_stmts(&self, env: &<D as Deploy<'a>>::CompileEnv) -> BTreeMap<usize, Vec<syn::Stmt>> {
        let mut extra_stmts: BTreeMap<usize, Vec<syn::Stmt>> = BTreeMap::new();
        for &c_id in self.clusters.keys() {
            let self_id_ident = syn::Ident::new(
                &format!("__hydro_lang_cluster_self_id_{}", c_id),
                Span::call_site(),
            );
            let self_id_expr = D::cluster_self_id(env).splice_untyped();
            extra_stmts
                .entry(c_id)
                .or_default()
                .push(syn::parse_quote! {
                    let #self_id_ident = #self_id_expr;
                });

            for other_location in self.processes.keys().chain(self.clusters.keys()) {
                let other_id_ident = syn::Ident::new(
                    &format!("__hydro_lang_cluster_ids_{}", c_id),
                    Span::call_site(),
                );
                let other_id_expr = D::cluster_ids(env, c_id).splice_untyped();
                extra_stmts
                    .entry(*other_location)
                    .or_default()
                    .push(syn::parse_quote! {
                        let #other_id_ident = #other_id_expr;
                    });
            }
        }
        extra_stmts
    }
}

impl<'a, D: Deploy<'a, CompileEnv = ()>> DeployFlow<'a, D> {
    #[must_use]
    pub fn deploy(mut self, env: &mut D::InstantiateEnv) -> DeployResult<'a, D> {
        self.used = true;

        let mut seen_tees_instantiate: HashMap<_, _> = HashMap::new();
        let mut seen_tee_locations: HashMap<_, _> = HashMap::new();
        self.ir.iter_mut().for_each(|leaf| {
            leaf.compile_network::<D>(
                &(),
                &mut seen_tees_instantiate,
                &mut seen_tee_locations,
                &self.processes,
                &self.clusters,
                &self.externals,
            );
        });

        let mut compiled = build_inner(&mut self.ir);
        let mut extra_stmts = self.extra_stmts(&());
        let mut meta = D::Meta::default();

        let (mut processes, mut clusters, mut externals) = (
            std::mem::take(&mut self.processes)
                .into_iter()
                .filter_map(|(node_id, node)| {
                    if let Some(ir) = compiled.remove(&node_id) {
                        node.instantiate(
                            env,
                            &mut meta,
                            ir,
                            extra_stmts.remove(&node_id).unwrap_or_default(),
                        );
                        Some((node_id, node))
                    } else {
                        None
                    }
                })
                .collect::<HashMap<_, _>>(),
            std::mem::take(&mut self.clusters)
                .into_iter()
                .filter_map(|(cluster_id, cluster)| {
                    if let Some(ir) = compiled.remove(&cluster_id) {
                        cluster.instantiate(
                            env,
                            &mut meta,
                            ir,
                            extra_stmts.remove(&cluster_id).unwrap_or_default(),
                        );
                        Some((cluster_id, cluster))
                    } else {
                        None
                    }
                })
                .collect::<HashMap<_, _>>(),
            std::mem::take(&mut self.externals)
                .into_iter()
                .map(|(external_id, external)| {
                    external.instantiate(
                        env,
                        &mut meta,
                        compiled.remove(&external_id).unwrap(),
                        extra_stmts.remove(&external_id).unwrap_or_default(),
                    );
                    (external_id, external)
                })
                .collect::<HashMap<_, _>>(),
        );

        for node in processes.values_mut() {
            node.update_meta(&meta);
        }

        for cluster in clusters.values_mut() {
            cluster.update_meta(&meta);
        }

        for external in externals.values_mut() {
            external.update_meta(&meta);
        }

        let mut seen_tees_connect = HashMap::new();
        self.ir.iter_mut().for_each(|leaf| {
            leaf.connect_network(&mut seen_tees_connect);
        });

        DeployResult {
            processes,
            clusters,
            externals,
            cluster_id_name: std::mem::take(&mut self.cluster_id_name)
                .into_iter()
                .collect(),
        }
    }
}

pub struct DeployResult<'a, D: Deploy<'a>> {
    processes: HashMap<usize, D::Process>,
    clusters: HashMap<usize, D::Cluster>,
    externals: HashMap<usize, D::ExternalProcess>,
    cluster_id_name: HashMap<usize, String>,
}

impl<'a, D: Deploy<'a>> DeployResult<'a, D> {
    pub fn get_process<P>(&self, p: &Process<P>) -> &D::Process {
        let id = match p.id() {
            LocationId::Process(id) => id,
            _ => panic!("Process ID expected"),
        };

        self.processes.get(&id).unwrap()
    }

    pub fn get_cluster<C>(&self, c: &Cluster<'a, C>) -> &D::Cluster {
        let id = match c.id() {
            LocationId::Cluster(id) => id,
            _ => panic!("Cluster ID expected"),
        };

        self.clusters.get(&id).unwrap()
    }

    pub fn get_all_clusters(&self) -> impl Iterator<Item = (LocationId, String, &D::Cluster)> {
        self.clusters.iter().map(|(&id, c)| {
            (
                LocationId::Cluster(id),
                self.cluster_id_name.get(&id).unwrap().clone(),
                c,
            )
        })
    }

    pub fn get_external<P>(&self, p: &ExternalProcess<P>) -> &D::ExternalProcess {
        self.externals.get(&p.id).unwrap()
    }

    pub fn raw_port(&self, port: ExternalBytesPort) -> D::ExternalRawPort {
        self.externals
            .get(&port.process_id)
            .unwrap()
            .raw_port(port.port_id)
    }

    pub async fn connect_sink_bytes(
        &self,
        port: ExternalBytesPort,
    ) -> Pin<Box<dyn Sink<Bytes, Error = Error>>> {
        self.externals
            .get(&port.process_id)
            .unwrap()
            .as_bytes_sink(port.port_id)
            .await
    }

    pub async fn connect_sink_bincode<T: Serialize + DeserializeOwned + 'static>(
        &self,
        port: ExternalBincodeSink<T>,
    ) -> Pin<Box<dyn Sink<T, Error = Error>>> {
        self.externals
            .get(&port.process_id)
            .unwrap()
            .as_bincode_sink(port.port_id)
            .await
    }

    pub async fn connect_source_bytes(
        &self,
        port: ExternalBytesPort,
    ) -> Pin<Box<dyn Stream<Item = Bytes>>> {
        self.externals
            .get(&port.process_id)
            .unwrap()
            .as_bytes_source(port.port_id)
            .await
    }

    pub async fn connect_source_bincode<T: Serialize + DeserializeOwned + 'static>(
        &self,
        port: ExternalBincodeStream<T>,
    ) -> Pin<Box<dyn Stream<Item = T>>> {
        self.externals
            .get(&port.process_id)
            .unwrap()
            .as_bincode_source(port.port_id)
            .await
    }
}
