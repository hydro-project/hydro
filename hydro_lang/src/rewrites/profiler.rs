use std::cell::RefCell;

use dfir_rs::futures::channel::mpsc::UnboundedSender;
use stageleft::*;

use super::profiler as myself; // TODO(shadaj): stageleft does not support `self::...`
use crate::ir::*;

pub fn increment_counter(count: &mut u64) {
    *count += 1;
}

fn quoted_any_fn<'a, F: Fn(&usize) + 'a, Q: IntoQuotedMut<'a, F, ()>>(q: Q) -> Q {
    q
}

/// Add a profiling node before each node to count the cardinality of its input
fn add_profiling_node<'a>(
    node: &mut HydroNode,
    counters: RuntimeData<&'a RefCell<Vec<u64>>>,
    counter_queue: RuntimeData<&'a RefCell<UnboundedSender<(usize, u64)>>>,
    my_id: usize,
) {
    let orig_node = std::mem::replace(node, HydroNode::Placeholder);
    let new_metadata = orig_node.metadata().clone();
    *node = HydroNode::Inspect {
        f: quoted_any_fn(q!({
            // Put counters on queue
            counter_queue
                .borrow()
                .unbounded_send((my_id, counters.borrow()[my_id]))
                .unwrap();
            counters.borrow_mut()[my_id] = 0;
            move |_| {
                myself::increment_counter(&mut counters.borrow_mut()[my_id]);
            }
        }))
        .splice_untyped()
        .into(),
        input: Box::new(orig_node),
        metadata: new_metadata,
    }
}

/// Count the cardinality of each input and periodically output to a file
#[cfg(feature = "build")]
#[stageleft::runtime]
pub fn profiling<'a>(
    ir: &mut Vec<HydroLeaf>,
    counters: RuntimeData<&'a RefCell<Vec<u64>>>,
    counter_queue: RuntimeData<&'a RefCell<UnboundedSender<(usize, u64)>>>,
) {
    traverse_dfir(ir, |_, _| (),
    |node, next_stmt_id| add_profiling_node(node, counters, counter_queue,next_stmt_id));
}

#[cfg(test)]
mod tests {
    use stageleft::*;

    use crate::deploy::MultiGraph;
    use crate::location::Location;

    #[test]
    fn profiler_wrapping_all_operators() {
        let flow = crate::builder::FlowBuilder::new();
        let process = flow.process::<()>();

        process
            .source_iter(q!(0..10))
            .map(q!(|v| v + 1))
            .for_each(q!(|n| println!("{}", n)));

        let built = flow.finalize();

        insta::assert_debug_snapshot!(&built.ir());

        // Print mermaid
        // let mut mermaid_config = WriteConfig {op_text_no_imports: true, ..Default::default()};
        // for (_, ir) in built.clone().with_default_optimize().compile().dfir() {
        //     println!("{}", ir.to_mermaid(&mermaid_config));
        // }

        let counters = RuntimeData::new("Fake");
        let counter_queue = RuntimeData::new("Fake");

        let pushed_down = built
            .optimize_with(crate::rewrites::persist_pullup::persist_pullup)
            .optimize_with(|ir| super::profiling(ir, counters, counter_queue));

        insta::assert_debug_snapshot!(&pushed_down.ir());

        let _ = pushed_down.compile_no_network::<MultiGraph>();
    }
}
