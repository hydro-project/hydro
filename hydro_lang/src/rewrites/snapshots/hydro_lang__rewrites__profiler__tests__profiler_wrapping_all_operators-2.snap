---
source: hydro_lang/src/rewrites/profiler.rs
expression: "&pushed_down.ir()"
---
[
    ForEach {
        f: stageleft :: runtime_support :: fn1_type_hint :: < i32 , () > ({ use crate :: __staged :: rewrites :: profiler :: tests :: * ; | n | println ! ("{}" , n) }),
        input: Inspect {
            f: { use crate :: __staged :: rewrites :: profiler :: * ; let counter_queue__free = Fake ; let counters__free = Fake ; let my_id__free = 1usize ; { counter_queue__free . borrow () . unbounded_send ((my_id__free , counters__free . borrow () [my_id__free])) . unwrap () ; counters__free . borrow_mut () [my_id__free] = 0 ; move | _ | { myself :: increment_counter (& mut counters__free . borrow_mut () [my_id__free]) ; } } },
            input: Map {
                f: stageleft :: runtime_support :: fn1_type_hint :: < i32 , i32 > ({ use crate :: __staged :: rewrites :: profiler :: tests :: * ; | v | v + 1 }),
                input: Inspect {
                    f: { use crate :: __staged :: rewrites :: profiler :: * ; let counter_queue__free = Fake ; let counters__free = Fake ; let my_id__free = 0usize ; { counter_queue__free . borrow () . unbounded_send ((my_id__free , counters__free . borrow () [my_id__free])) . unwrap () ; counters__free . borrow_mut () [my_id__free] = 0 ; move | _ | { myself :: increment_counter (& mut counters__free . borrow_mut () [my_id__free]) ; } } },
                    input: Source {
                        source: Iter(
                            { use crate :: __staged :: rewrites :: profiler :: tests :: * ; 0 .. 10 },
                        ),
                        location_kind: Process(
                            0,
                        ),
                        metadata: HydroIrMetadata {
                            location_kind: Process(
                                0,
                            ),
                            output_type: Some(
                                i32,
                            ),
                        },
                    },
                    metadata: HydroIrMetadata {
                        location_kind: Process(
                            0,
                        ),
                        output_type: Some(
                            i32,
                        ),
                    },
                },
                metadata: HydroIrMetadata {
                    location_kind: Process(
                        0,
                    ),
                    output_type: Some(
                        i32,
                    ),
                },
            },
            metadata: HydroIrMetadata {
                location_kind: Process(
                    0,
                ),
                output_type: Some(
                    i32,
                ),
            },
        },
    },
]
